Add hacked metric to watchtower report

andriyun opened the pull request:
Hacked check process based on unix diff tool
1. Original project download to /tmp/dwt_hacked folder and rename to version related name
2. run unix diff tool and parse output using diffstat tool

andriyun committed:
Based implementation of hacked check for contrib modules

andriyun committed:
Add hacked check for drupal core

jbguerraz commented on dw_client/watchtower.drush.inc:
```diff
@@ -222,6 +227,124 @@ function watchtower_search_patches($modules, $recursive = FALSE) {
+
+        // @todo Drush outputs to err somehow.
+        if (trim($output) == '0 files changed') {
+          $hacked_result = "[OK] Project code mached";
```
matched not mached.

andypost commented on dw_client/watchtower.drush.inc:
```diff
@@ -222,6 +227,124 @@ function watchtower_search_patches($modules, $recursive = FALSE) {
+        2 => array('pipe', 'w'),
+      );
+      $pipes = array();
+      if ($h = proc_open($command, $descriptors, $pipes, $root)) {
```
Better wrap into try... Catch

andypost commented on dw_client/watchtower.drush.inc:
```diff
@@ -222,6 +227,124 @@ function watchtower_search_patches($modules, $recursive = FALSE) {
+        2 => array('pipe', 'w'),
+      );
+      $pipes = array();
+      if ($h = proc_open($command, $descriptors, $pipes, $root)) {
```
The same (wrap into try / catch)

jbguerraz commented on dw_client/watchtower.drush.inc:
```diff
@@ -222,6 +227,124 @@ function watchtower_search_patches($modules, $recursive = FALSE) {
+
+      $exclude = '';
+      if ($module_name == 'drupal') {
+        $exclude = '--exclude=sites ';
```
have to exclude custom profiles as well (but not core profiles)

jbguerraz commented on dw_client/watchtower.drush.inc:
```diff
@@ -222,6 +227,124 @@ function watchtower_search_patches($modules, $recursive = FALSE) {
+ */
+function watchtower_check_hacked($modules, $recursive = FALSE) {
+  $hacked = array();
+  $tmp_dir = '/tmp/dwt_hacked';
```
what about file_directory_temp() ? why not to use it ?

jbguerraz commented on dw_client/watchtower.drush.inc:
```diff
@@ -175,6 +175,11 @@ function watchtower_get_default_report() {
   }
 
+  // Add Hacked.
+  if ($hacked = watchtower_check_hacked($modules)) {
```
watchtower_check_hacked() always returns an array no ? if so, not sure that "if" is needed. we could return the hacked array only if the count($check_changed) is > 0 OR $check_counter != count($hacked) (to return the notice if some projects was not downloadable) and returns false otherwhise and then that if would be needed.

jbguerraz commented:
I believe we need changes details in report.

Also a global 2 cents feedback : beware strings mapping => http://stackoverflow.com/a/3446286

jbguerraz commented on dw_client/watchtower.drush.inc:
```diff
@@ -222,6 +227,124 @@ function watchtower_search_patches($modules, $recursive = FALSE) {
+          $check_ok++;
+        }
+        else {
+          $hacked_result = "[WARNING] Project " . $module_name ." was changed\n";
```
2 cents feedback : why " and not ' ? :)

jbguerraz commented on dw_client/watchtower.drush.inc:
```diff
@@ -222,6 +227,124 @@ function watchtower_search_patches($modules, $recursive = FALSE) {
+    }
+    $hacked[$module_name] = $hacked_result;
+  }
+  drush_print("Checked " . $check_counter . ' from ' . count($hacked) . ' modules.');
```
2 cents feedback : why " and not ' ? :)

andriyun committed:
Fix quotes mess. Few small adjustments

andriyun committed:
Exclude custom profiles

andriyun committed:
We use drupal tmp folder

andriyun committed:
Change checking strategy. For now we check only existing files.

andriyun committed:
Optimize report data

andriyun committed:
Rewrite proc_open command using by try catch

andriyun committed:
Add process exitcode support with print error message.

andriyun committed:
Rewrite command call with using process success flag. Add adjustments to report info.

andriyun commented:
- added try/catch for common process call
- added exitcode support of process call
- changed hacked report structure - we returned only hacked info

Need review again

andriyun committed:
Changed file_scan dir to Hacked! version for get hidden files

andriyun committed:
Add exists condition for project download step

andriyun committed:
Adjustments for module download condition

andriyun committed:
Rewrite drush_print to drush_log

andriyun committed:
Fix undefined variable
